edx_arch_experiments/datadog_monitoring/scripts/datadog_search.py

```diff
@@ -0,0 +1,186 @@
+"""
+This script takes a regex to search through the Datadog monitors
+and dashboards.`
+
+For help::
+
+    python edx_arch_experiments/datadog_monitoring/scripts/datadog_search.py --help
+
+"""
+import datetime
+import re
+import types
+
+import click
+from datadog_api_client import ApiClient, Configuration
+from datadog_api_client.v1.api.dashboards_api import DashboardsApi
+from datadog_api_client.v1.api.monitors_api import MonitorsApi
+
+
+@click.command()
+@click.option(
+    '--regex',
+    required=True,
+    help="The regex to use to search in monitors and dashboards.",
+)
+def main(regex):
+    """
+    Search Datadog monitors and dashboards using regex.
+
+    Example usage:
+
+        python datadog_search.py --regex tnl
+
+    Note: The search ignores case since most features are case insensitive.
+
+    Pre-requisites:
+    1. Install the client library:
+        pip install datadog-api-client
+    2. Set the following environment variables (in a safe way):
+        export DD_API_KEY=XXXXX
+        export DD_APP_KEY=XXXXX
+    See https://docs.datadoghq.com/api/latest/?code-lang=python for more details.
+
+    If you get a Forbidden error, you either didn't supply a proper DD_API_KEY and
+    DD_APP_KEY, or your DD_APP_KEY is missing certain required scopes:
+    - dashboards_read
+    - monitors_read
+
+    For developing with the datadog-api-client, see:
+    - https://github.com/DataDog/datadog-api-client-python
+
+    """
+    compiled_regex = re.compile(regex)
+    configuration = Configuration()
+    api_client = ApiClient(configuration)
+
+    search_monitors(compiled_regex, api_client)
+    print('\n')
+    search_dashboards(compiled_regex, api_client)
+    print(flush=True)
+
+
+def search_monitors(regex, api_client):
+    """
+    Searches Datadog monitors using the regex argument.
+
+    Arguments:
+        regex (re.Pattern): compiled regex used to find matches.
+        api_client (int): a Datadog client for making API requests.
+    """
+    api_instance = MonitorsApi(api_client)
+
+    print(f"Searching for regex {regex.pattern} in all monitors:")
+    match_found = False
+    for monitor in api_instance.list_monitors_with_pagination():
+        matches = find_matches(regex, monitor, 'monitor')
+        if matches:
+            print('\n')
+            print(f'- {monitor.id} "{monitor.name}" {monitor.tags}')
+            for match in matches:
+                print(f'  - query_path: {match[1]}')
```

Comment:
What benefit does the query_path have -- is it to help identify which widget in a dashboard matched? (Are the results in the query results even ordered the same as they are in the UI?)

Reply:
If we didn't have to print the query_path, then find_matches could _probably_ be reduced to a repr + re.search call.

Reply:
- I like that it gives some indication of where to look (e.g. title, vs widget query, vs widget text, etc.). If you searched for something in a facet or group, this would also be indicated. It's imperfect, but gives hints.
- The NRQL search used to mostly just search NRQL queries. This script started out searching things with name `query`, but it was more complicated and less useful. I've updated my outdated language of `query` and `query_path` to `match` and `match_path`, which I had missed in my earlier pass. Thanks.

Reply:
Makes sense!